File: ios/Classes/CallAudioSession.swift
import AVFAudio
import WebRTC

/// Manual-mode control of WebRTC's audio session.
///
/// CallKit owns `AVAudioSession` for any call it presents: it activates the
/// session and tells the app afterwards. WebRTC must therefore be in manual
/// mode, or the two race and the call ends up with no audio while every other
/// signal — tracks published, peer connection connected — looks healthy.
///
/// The notifications that drive this arrive through the host `AppDelegate`;
/// see `CallNativeKitHost`. If that forwarding is missing, [awaitActive] times
/// out and the call is silent, which is why [diagnostics] exists.
public enum CallAudioSession {

    private static let lock = NSLock()
    private static var nextWaiterToken = 0
    private static var pendingWaiters: [Int: (Bool) -> Void] = [:]
    private static var activationCount = 0
    private static var lastActivatedAt: Date?

    /// Hands the session to CallKit. Call before any call can arrive.
    public static func configureManualMode() {
        RTCAudioSession.sharedInstance().useManualAudio = true
        RTCAudioSession.sharedInstance().isAudioEnabled = false
    }

    /// Forward from `CallkitIncomingAppDelegate.didActivateAudioSession`.
    public static func didActivate(_ audioSession: AVAudioSession) {
        RTCAudioSession.sharedInstance().audioSessionDidActivate(audioSession)
        RTCAudioSession.sharedInstance().isAudioEnabled = true
        lock.lock()
        activationCount += 1
        lastActivatedAt = Date()
        lock.unlock()
        signalWaiters(activated: true)
    }

    /// Forward from `CallkitIncomingAppDelegate.didDeactivateAudioSession`.
    public static func didDeactivate(_ audioSession: AVAudioSession) {
        RTCAudioSession.sharedInstance().audioSessionDidDeactivate(audioSession)
        RTCAudioSession.sharedInstance().isAudioEnabled = false
    }

    /// Activates the session for a call that never went through CallKit —
    /// an outgoing call, or joining a room from inside the app.
    ///
    /// Idempotent: if CallKit already activated the session, this only
    /// re-asserts the routing.
    public static func activateForCall(defaultToSpeaker: Bool) {
        let session = AVAudioSession.sharedInstance()
        var options: AVAudioSession.CategoryOptions = [
            .allowBluetooth, .allowBluetoothA2DP,
        ]
        if defaultToSpeaker { options.insert(.defaultToSpeaker) }
        do {
            try session.setCategory(.playAndRecord, mode: .voiceChat, options: options)
            try session.setActive(true, options: .notifyOthersOnDeactivation)
        } catch {
            NSLog("[call_native_kit] activateForCall failed: \(error)")
        }
        RTCAudioSession.sharedInstance().audioSessionDidActivate(session)
        RTCAudioSession.sharedInstance().isAudioEnabled = true
        lock.lock()
        activationCount += 1
        lastActivatedAt = Date()
        lock.unlock()
    }

    public static func deactivateForCall() {
        let session = AVAudioSession.sharedInstance()
        RTCAudioSession.sharedInstance().isAudioEnabled = false
        RTCAudioSession.sharedInstance().audioSessionDidDeactivate(session)
        do {
            try session.setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            NSLog("[call_native_kit] deactivateForCall failed: \(error)")
        }
    }

    /// Waits for CallKit to activate the session, for a call the user accepted
    /// on the system UI. Completes exactly once, possibly off the main queue.
    public static func awaitActive(
        timeoutMs: Int,
        completion: @escaping (Bool) -> Void
    ) {
        if RTCAudioSession.sharedInstance().isAudioEnabled {
            completion(true)
            return
        }

        lock.lock()
        nextWaiterToken += 1
        let token = nextWaiterToken
        pendingWaiters[token] = completion
        lock.unlock()

        // Activation could have landed between the check above and the lock,
        // in which case signalWaiters ran against an empty table and this
        // waiter would otherwise sit until the timeout for no reason.
        if RTCAudioSession.sharedInstance().isAudioEnabled {
            lock.lock()
            let waiter = pendingWaiters.removeValue(forKey: token)
            lock.unlock()
            waiter?(true)
            return
        }

        DispatchQueue.global().asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) {
            lock.lock()
            let waiter = pendingWaiters.removeValue(forKey: token)
            lock.unlock()
            waiter?(false)
        }
    }

    /// Reported to Dart so a silent call is visible rather than merely silent.
    public static func diagnostics() -> [String: Any] {
        lock.lock()
        let count = activationCount
        let last = lastActivatedAt
        lock.unlock()

        let session = AVAudioSession.sharedInstance()
        var payload: [String: Any] = [
            "isActive": RTCAudioSession.sharedInstance().isActive,
            "isAudioEnabled": RTCAudioSession.sharedInstance().isAudioEnabled,
            "activationCount": count,
            "category": session.category.rawValue,
            "mode": session.mode.rawValue,
        ]
        if let last {
            payload["lastActivatedAtMs"] = Int(last.timeIntervalSince1970 * 1000)
        }
        return payload
    }

    private static func signalWaiters(activated: Bool) {
        lock.lock()
        let waiters = Array(pendingWaiters.values)
        pendingWaiters.removeAll()
        lock.unlock()
        for waiter in waiters { waiter(activated) }
    }
}

File: packages/call_native_kit/lib/src/audio/call_audio_session.dart
import 'dart:io';

import '../channels.dart';
import '../config/call_native_timeouts.dart';
import '../logging/call_logger.dart';

/// State of the WebRTC audio session, for diagnosing calls that connect
/// perfectly and carry no sound.
class AudioSessionDiagnostics {
  const AudioSessionDiagnostics({
    required this.isActive,
    required this.isAudioEnabled,
    required this.activationCount,
    required this.category,
    required this.mode,
    this.lastActivatedAt,
  });

  final bool isActive;
  final bool isAudioEnabled;

  /// How many times the system activated the session this process. Zero
  /// during a CallKit-accepted call means the delegate chain is broken.
  final int activationCount;

  final String category;
  final String mode;
  final DateTime? lastActivatedAt;

  static AudioSessionDiagnostics fromJson(Map<String, Object?> json) {
    final lastMs = json['lastActivatedAtMs'];
    return AudioSessionDiagnostics(
      isActive: json['isActive'] as bool? ?? false,
      isAudioEnabled: json['isAudioEnabled'] as bool? ?? false,
      activationCount: json['activationCount'] as int? ?? 0,
      category: json['category'] as String? ?? '',
      mode: json['mode'] as String? ?? '',
      lastActivatedAt:
          lastMs is int ? DateTime.fromMillisecondsSinceEpoch(lastMs) : null,
    );
  }

  @override
  String toString() =>
      'AudioSessionDiagnostics(active: $isActive, audioEnabled: '
      '$isAudioEnabled, activations: $activationCount, category: $category, '
      'mode: $mode, lastActivatedAt: $lastActivatedAt)';
}

/// Manual-mode control of the WebRTC audio session.
///
/// ## Why this is not automatic
///
/// On iOS, CallKit — not the app — owns `AVAudioSession` for a call it
/// presents. It activates the session and tells the app afterwards. If the
/// app also activates the session, the two fight and the call ends up with no
/// audio at all: everything else looks healthy, tracks publish, the peer
/// connection reports connected, and nobody hears anything.
///
/// So the rule is:
///
/// * call accepted through the system UI → [awaitActive], never [activate];
/// * call started from inside the app → [activate].
///
/// The chain that makes [awaitActive] complete runs through the host
/// `AppDelegate`: it must conform to `CallkitIncomingAppDelegate` and forward
/// `didActivateAudioSession`. If that forwarding is missing, [awaitActive]
/// times out and the call is silent — which is why the timeout is reported
/// through [CallLogger.recordError] rather than merely logged.
class CallAudioSession {
  CallAudioSession({
    required CallNativeTimeouts Function() timeouts,
    CallLogger logger = const SilentCallLogger(),
  })  : _timeouts = timeouts,
        _logger = logger;

  final CallNativeTimeouts Function() _timeouts;
  final CallLogger _logger;

  /// Activates the session for a call the app started itself.
  Future<void> activate({bool defaultToSpeaker = true}) async {
    try {
      await CallNativeChannels.main
          .invokeMethod<void>(CallNativeMethods.activateAudio, {
        'defaultToSpeaker': defaultToSpeaker,
      });
    } catch (e, st) {
      _logger.recordError(e, st, reason: 'activate audio session');
    }
  }

  Future<void> deactivate() async {
    try {
      await CallNativeChannels.main.invokeMethod<void>(
        CallNativeMethods.deactivateAudio,
      );
    } catch (e, st) {
      _logger.recordError(e, st, reason: 'deactivate audio session');
    }
  }

  /// Waits for the system to activate the session, for a call accepted
  /// through the system UI.
  ///
  /// Returns `false` on timeout. A `false` here is the signature of a silent
  /// call — see the class doc.
  Future<bool> awaitActive({Duration? timeout}) async {
    if (!Platform.isIOS) return true;
    final wait = timeout ?? _timeouts().audioSessionWait;
    try {
      final active = await CallNativeChannels.main
              .invokeMethod<bool>(CallNativeMethods.awaitAudioSessionActive, {
            'timeoutMs': wait.inMilliseconds,
          }) ??
          false;
      if (!active) {
        _logger.recordError(
          StateError('audio session not active after ${wait.inMilliseconds}ms'),
          StackTrace.current,
          reason:
              'CallKit did not activate AVAudioSession — check that the host '
              'AppDelegate conforms to CallkitIncomingAppDelegate and '
              'forwards didActivateAudioSession',
        );
      }
      return active;
    } catch (e, st) {
      _logger.recordError(e, st, reason: 'awaitAudioSessionActive');
      return false;
    }
  }

  Future<AudioSessionDiagnostics?> diagnostics() async {
    try {
      final raw = await CallNativeChannels.main
          .invokeMapMethod<String, Object?>(CallNativeMethods.audioDiagnostics);
      if (raw == null) return null;
      return AudioSessionDiagnostics.fromJson(raw);
    } catch (e, st) {
      _logger.recordError(e, st, reason: 'audioDiagnostics');
      return null;
    }
  }
}
